// @practica/test-eyes/dist/file-operations.js
import fs from 'fs/promises';
import path from 'path';
export async function ensureDir(dirPath) {
    await fs.mkdir(dirPath, { recursive: true });
}
export function generateRunId(commitSha) {
    const date = new Date().toISOString().split('T')[0];
    const shortSha = commitSha.slice(0, 7);
    return `${date}_${shortSha}`;
}
export function generateFilename(commitSha) {
    return `${generateRunId(commitSha)}.json`;
}
export async function saveRunData(dataDir, filename, data) {
    await ensureDir(dataDir);
    const filepath = path.join(dataDir, filename);
    await fs.writeFile(filepath, JSON.stringify(data, null, 2));
    return filepath;
}
export async function loadJson(filepath) {
    try {
        const content = await fs.readFile(filepath, 'utf-8');
        return JSON.parse(content);
    }
    catch {
        return null;
    }
}
export async function saveJson(filepath, data) {
    await fs.writeFile(filepath, JSON.stringify(data, null, 2));
}
export async function fileExists(filepath) {
    try {
        await fs.access(filepath);
        return true;
    }
    catch {
        return false;
    }
}
export async function listJsonFiles(dirPath) {
    try {
        const files = await fs.readdir(dirPath);
        return files.filter((f) => f.endsWith('.json'));
    }
    catch {
        return [];
    }
}
