// @practica/test-eyes/dist/git-operations.js
import { exec } from 'child_process';
import { promisify } from 'util';
const execAsync = promisify(exec);
const MAX_BUFFER = 50 * 1024 * 1024; // 50MB
async function run(command) {
    const { stdout } = await execAsync(command, { maxBuffer: MAX_BUFFER });
    return stdout.trim();
}
export function getDefaultGitConfig() {
    return {
        userName: 'github-actions[bot]',
        userEmail: 'github-actions[bot]@users.noreply.github.com'
    };
}
export async function configureGit(config) {
    await run(`git config user.name "${config.userName}"`);
    await run(`git config user.email "${config.userEmail}"`);
}
export async function getCurrentSha() {
    return run('git rev-parse HEAD');
}
export async function getCurrentBranch() {
    return run('git rev-parse --abbrev-ref HEAD');
}
export async function fetchBranch(branch) {
    try {
        await run(`git fetch origin ${branch}:${branch}`);
        return true;
    }
    catch {
        return false;
    }
}
export async function branchExists(branch) {
    try {
        await run(`git rev-parse --verify ${branch}`);
        return true;
    }
    catch {
        return false;
    }
}
export async function checkoutBranch(branch) {
    await run(`git checkout ${branch}`);
}
export async function createOrphanBranch(branch) {
    await run(`git checkout --orphan ${branch}`);
    await run('git rm -rf . || true');
}
export async function stageFiles(patterns) {
    for (const pattern of patterns) {
        await run(`git add ${pattern}`);
    }
}
export async function hasChanges() {
    try {
        const result = await run('git status --porcelain');
        return result.length > 0;
    }
    catch {
        return false;
    }
}
export async function commit(message) {
    try {
        await run(`git commit --no-verify -m "${message}"`);
        return await getCurrentSha();
    }
    catch {
        return null;
    }
}
export async function push(branch) {
    try {
        await run(`git push origin ${branch} --force`);
        return true;
    }
    catch {
        return false;
    }
}
