// @practica/test-eyes/dist/junit-parser.js
import { readFile } from 'fs/promises';
function normalizeToArray(value) {
    if (!value)
        return [];
    return Array.isArray(value) ? value : [value];
}
function getTestStatus(testCase) {
    if (testCase.skipped !== undefined)
        return 'skipped';
    if (testCase.failure !== undefined || testCase.error !== undefined)
        return 'failed';
    return 'passed';
}
function parseTestCase(testCase) {
    const className = testCase['@_classname'] || '';
    const testName = testCase['@_name'] || 'unknown';
    const name = className ? `${className} ${testName}` : testName;
    return {
        name: name.trim(),
        durationMs: Math.round(parseFloat(testCase['@_time'] || '0') * 1000),
        status: getTestStatus(testCase)
    };
}
export function parseJUnitXml(xml) {
    // Simple XML parser - no external dependencies
    const tests = [];
    // Extract all testcase elements
    const testCaseRegex = /<testcase([^>]*)(?:\/>|>([\s\S]*?)<\/testcase>)/g;
    let match;
    while ((match = testCaseRegex.exec(xml)) !== null) {
        const attrs = match[1];
        const content = match[2] || '';
        const nameMatch = attrs.match(/name=["']([^"']+)["']/);
        const classnameMatch = attrs.match(/classname=["']([^"']+)["']/);
        const timeMatch = attrs.match(/time=["']([^"']+)["']/);
        const name = nameMatch?.[1] || 'unknown';
        const classname = classnameMatch?.[1] || '';
        const time = timeMatch?.[1] || '0';
        let status = 'passed';
        if (content.includes('<skipped') || attrs.includes('skipped')) {
            status = 'skipped';
        }
        else if (content.includes('<failure') || content.includes('<error')) {
            status = 'failed';
        }
        tests.push({
            name: classname ? `${classname} ${name}` : name,
            durationMs: Math.round(parseFloat(time) * 1000),
            status
        });
    }
    return tests;
}
export async function parseJUnitFile(filepath) {
    const xml = await readFile(filepath, 'utf-8');
    return parseJUnitXml(xml);
}
