// @practica/test-eyes/dist/reporter.js
import { generateRunId, generateFilename, saveRunData, ensureDir } from './file-operations.js';
import { configureGit, getDefaultGitConfig, getCurrentSha, getCurrentBranch, fetchBranch, branchExists, checkoutBranch, createOrphanBranch, stageFiles, hasChanges, commit, push } from './git-operations.js';
import { aggregate } from './aggregate.js';
import { deployDashboard } from './deploy.js';
export class TestEyesReporter {
    options;
    tests = [];
    originalBranch = '';
    startTime = 0;
    constructor(options = {}) {
        this.options = {
            dataBranch: options.dataBranch ?? 'gh-data',
            deployBranch: options.deployBranch ?? 'gh-pages',
            deploy: options.deploy ?? true,
            frontendDistPath: options.frontendDistPath ?? '',
            prNumber: options.prNumber ?? this.getPrNumber()
        };
    }
    getPrNumber() {
        return parseInt(process.env.GITHUB_PR_NUMBER || process.env.PR_NUMBER || '0', 10) || 0;
    }
    isCI() {
        return !!(process.env.CI || process.env.GITHUB_ACTIONS);
    }
    log(message) {
        console.log(`[test-eyes] ${message}`);
    }
    logError(message, error) {
        const errorMsg = error instanceof Error ? error.message : String(error ?? '');
        console.error(`[test-eyes] ERROR: ${errorMsg ? `${message}: ${errorMsg}` : message}`);
    }
    onBegin(_config, _suite) {
        this.startTime = Date.now();
        this.tests = [];
        this.log('Starting test collection...');
    }
    onTestEnd(test, result) {
        this.tests.push({
            name: test.titlePath().filter(Boolean).join(' > '),
            durationMs: result.duration,
            status: this.mapStatus(result.status),
            wasFlaky: result.status === 'passed' && result.retry > 0
        });
    }
    mapStatus(status) {
        if (status === 'passed')
            return 'passed';
        if (status === 'skipped')
            return 'skipped';
        return 'failed';
    }
    async onEnd(_result) {
        this.log(`Tests completed in ${Date.now() - this.startTime}ms`);
        this.log(`Collected ${this.tests.length} test results`);
        const flakyCount = this.tests.filter(t => t.wasFlaky).length;
        if (flakyCount > 0)
            this.log(`Detected ${flakyCount} flaky tests`);
        if (!this.isCI()) {
            this.log('Not in CI, skipping. Set CI=true to enable.');
            return;
        }
        await this.submitData();
    }
    async submitData() {
        const context = await this.prepareContext();
        if (!context)
            return;
        const success = await this.saveToDataBranch(context);
        await this.returnToOriginalBranch(context.originalBranch);
        if (success && this.options.deploy) {
            await this.deploy();
        }
    }
    async prepareContext() {
        try {
            await configureGit(getDefaultGitConfig());
            const commitSha = process.env.GITHUB_SHA || await getCurrentSha();
            const originalBranch = await getCurrentBranch();
            return {
                commitSha,
                originalBranch,
                runData: {
                    runId: generateRunId(commitSha),
                    prNumber: this.options.prNumber,
                    commitSha,
                    createdAt: new Date().toISOString(),
                    tests: this.tests
                }
            };
        }
        catch (error) {
            this.logError('Failed to prepare context', error);
            return null;
        }
    }
    async saveToDataBranch(ctx) {
        const { dataBranch } = this.options;
        try {
            await this.checkoutDataBranch(dataBranch);
            await this.saveData(ctx);
            await this.commitAndPush(ctx.runData.runId, dataBranch);
            return true;
        }
        catch (error) {
            this.logError('Failed to save data', error);
            return false;
        }
    }
    async checkoutDataBranch(branch) {
        const exists = await fetchBranch(branch) || await branchExists(branch);
        if (exists) {
            await checkoutBranch(branch);
        }
        else {
            this.log(`Creating new branch: ${branch}`);
            await createOrphanBranch(branch);
        }
    }
    async saveData(ctx) {
        const dataDir = 'data';
        await ensureDir(dataDir);
        await saveRunData(dataDir, generateFilename(ctx.commitSha), ctx.runData);
        this.log(`Saved: ${dataDir}/${generateFilename(ctx.commitSha)}`);
        await aggregate(dataDir);
        this.log('Aggregation complete');
    }
    async commitAndPush(runId, branch) {
        await stageFiles(['data/']);
        if (!await hasChanges()) {
            this.log('No changes to commit');
            return;
        }
        const sha = await commit(`Add test data: ${runId}`);
        if (sha)
            this.log(`Committed: ${sha.slice(0, 7)}`);
        const pushed = await push(branch);
        if (pushed)
            this.log(`Pushed to ${branch}`);
    }
    async returnToOriginalBranch(branch) {
        if (branch && branch !== this.options.dataBranch) {
            try {
                await checkoutBranch(branch);
            }
            catch (error) {
                this.logError('Failed to return to original branch', error);
            }
        }
    }
    async deploy() {
        try {
            await deployDashboard({
                dataBranch: this.options.dataBranch,
                deployBranch: this.options.deployBranch
            });
        }
        catch (error) {
            this.logError('Deploy failed', error);
        }
    }
}
